File: flextarget/View/ImageCrop/ImageCropViewModel.swift
import SwiftUI
import Combine
#if canImport(UIKit)
import UIKit
#endif

class ImageCropViewModel: ObservableObject {
    @Published var selectedImage: UIImage?
    @Published var croppedImage: UIImage?
    @Published var scale: CGFloat = 1.0
    @Published var offset: CGSize = .zero
    @Published var showImagePicker = false
    @Published var showLivePreview = false
    
    // Calculated properties
    var minScale: CGFloat = 1.0
    var maxScale: CGFloat = 5.0
    
    func resetTransform() {
        scale = 1.0
        offset = .zero
    }

    /// Enforce minimum scale so the image always fills the provided cropSize
    /// and clamp offset so the crop rectangle remains covered by the image.
    func enforceConstraints(containerSize: CGSize, cropSize: CGSize) {
        // Ensure the user-visible image (container frame scaled by `scale`) fully
        // covers the crop rectangle `cropSize`. Adjust `scale` if it's too small
        // and clamp the `offset` so the crop rect never reveals background.

        // Compute the minimum user scale needed so the displayed image covers the crop area.
        // displayed size = containerSize * scale, so requiredScale = cropSize / containerSize
        let requiredScaleX = cropSize.width / max(0.0001, containerSize.width)
        let requiredScaleY = cropSize.height / max(0.0001, containerSize.height)
        let minAllowedScale = max(1.0, requiredScaleX, requiredScaleY)

        if scale < minAllowedScale {
            scale = minAllowedScale
        }

        // Compute the displayed image size (points) after applying base fill scale and user scale.
        // If we have the image, calculate how the image is scaled to fill the container (scaledToFill semantics):
        //   baseFillScale = max(containerWidth / imageWidth, containerHeight / imageHeight)
        // Displayed image size = image.size * baseFillScale * userScale
        let (displayedWidth, displayedHeight): (CGFloat, CGFloat) = {
            if let img = selectedImage {
                let baseFill = max(containerSize.width / max(0.0001, img.size.width),
                                   containerSize.height / max(0.0001, img.size.height))
                return (img.size.width * baseFill * scale, img.size.height * baseFill * scale)
            } else {
                // Fallback to container-based calculation if no image is available
                return (containerSize.width * scale, containerSize.height * scale)
            }
        }()

        // Allowed half-range for the center offset so the crop rect stays fully covered.
        let allowedHalfX = max(0, (displayedWidth - cropSize.width) / 2.0)
        let allowedHalfY = max(0, (displayedHeight - cropSize.height) / 2.0)

        func clamp(_ value: CGFloat, _ lower: CGFloat, _ upper: CGFloat) -> CGFloat {
            return Swift.min(Swift.max(value, lower), upper)
        }

        let clampedX = clamp(offset.width, -allowedHalfX, allowedHalfX)
        let clampedY = clamp(offset.height, -allowedHalfY, allowedHalfY)

        offset = CGSize(width: clampedX, height: clampedY)
    }

    /// Return a clamped offset for a proposed offset without mutating state.
    func clampedOffset(for proposed: CGSize, containerSize: CGSize, cropSize: CGSize, scaleOverride: CGFloat? = nil) -> CGSize {
        // Allow clamping even when `selectedImage` is nil; the math doesn't need the UIImage itself.
        let effectiveScale = scaleOverride ?? scale

        // Displayed image size (points) = container frame size * effective scale
        // Compute displayed image size (points) using image intrinsic size when possible
        let (displayedWidth, displayedHeight): (CGFloat, CGFloat) = {
            if let img = selectedImage {
                let baseFill = max(containerSize.width / max(0.0001, img.size.width),
                                   containerSize.height / max(0.0001, img.size.height))
                return (img.size.width * baseFill * effectiveScale, img.size.height * baseFill * effectiveScale)
            } else {
                return (containerSize.width * effectiveScale, containerSize.height * effectiveScale)
            }
        }()

        // Allowed half-range for the center offset so the crop rect remains fully covered
        let allowedHalfX = max(0, (displayedWidth - cropSize.width) / 2.0)
        let allowedHalfY = max(0, (displayedHeight - cropSize.height) / 2.0)

        // Helper to clamp a single value into [lower, upper]
        func clamp(_ value: CGFloat, _ lower: CGFloat, _ upper: CGFloat) -> CGFloat {
            return Swift.min(Swift.max(value, lower), upper)
        }

        let clampedX = clamp(proposed.width, -allowedHalfX, allowedHalfX)
        let clampedY = clamp(proposed.height, -allowedHalfY, allowedHalfY)

        return CGSize(width: clampedX, height: clampedY)
    }
    
    func cropImage(within cropFrame: CGRect, canvasSize: CGSize) {
        // Normalize the source image first so orientation is upright and size matches its pixel data
        #if canImport(UIKit)
        guard let src = selectedImage else { return }
        let norm = normalizedImage(src)
        guard let cg = norm.cgImage else { return }

        // Compute baseFillScale using the normalized image size
        let baseFill = max(canvasSize.width / max(0.0001, norm.size.width),
                           canvasSize.height / max(0.0001, norm.size.height))

        // The displayed image size in points after baseFill and user scale
        let displayedImageSize = CGSize(width: norm.size.width * baseFill * scale,
                                        height: norm.size.height * baseFill * scale)

        // Top-left of the displayed image inside the container (points)
        let imageOrigin = CGPoint(x: (canvasSize.width - displayedImageSize.width) / 2.0 + offset.width,
                                  y: (canvasSize.height - displayedImageSize.height) / 2.0 + offset.height)

        // Convert cropFrame (in container points) to normalized image points: (point - imageOrigin) / (baseFill * scale)
        let invScale = 1.0 / (baseFill * scale)

        let imgX = (cropFrame.minX - imageOrigin.x) * invScale
        let imgY = (cropFrame.minY - imageOrigin.y) * invScale
        let imgW = cropFrame.width * invScale
        let imgH = cropFrame.height * invScale

        // Convert image points to image pixels (cgImage coordinate) using normalized image.scale
        let pxPerPoint = norm.scale
        let origin = CGPoint(x: CGFloat(imgX * pxPerPoint), y: CGFloat(imgY * pxPerPoint))
        let size = CGSize(width: CGFloat(imgW * pxPerPoint), height: CGFloat(imgH * pxPerPoint))
        var cropRectPixels = CGRect(origin: origin, size: size)

        // Normalize to integer pixel boundaries and clamp to image bounds to avoid
        // fractional/out-of-range rectangles that can cause cg.cropping(to:) to fail
        cropRectPixels = cropRectPixels.standardized.integral
        let imageBounds = CGRect(x: 0, y: 0, width: CGFloat(cg.width), height: CGFloat(cg.height))
        cropRectPixels = cropRectPixels.intersection(imageBounds)

        // If the resulting rect is empty or too small, attempt a safe fallback
        if !(cropRectPixels.width >= 1.0 && cropRectPixels.height >= 1.0) {
            // Log diagnostics for the originally computed rect
            print("⚠️ cropImage: empty crop rect after clamping: \(cropRectPixels) imageBounds=\(imageBounds) img.size=\(norm.size) baseFill=\(baseFill) scale=\(scale) pxPerPoint=\(pxPerPoint)")

            // Fallback: create a centered crop that preserves the target aspect ratio
            // (720x1280) inside the available image bounds. This ensures a valid crop
            // rectangle even when the guide/cropFrame calculation produced a zero width/height.
            let targetSize = CGSize(width: 720.0, height: 1280.0)
            let targetAspect = targetSize.width / targetSize.height

            // Start by trying to use full image height and compute width by aspect
            var fallbackHeight = imageBounds.height
            var fallbackWidth = fallbackHeight * targetAspect

            // If that width exceeds bounds, use full width and compute height by aspect
            if fallbackWidth > imageBounds.width {
                fallbackWidth = imageBounds.width
                fallbackHeight = fallbackWidth / targetAspect
            }

            let fx = (imageBounds.width - fallbackWidth) / 2.0
            let fy = (imageBounds.height - fallbackHeight) / 2.0
            cropRectPixels = CGRect(x: fx, y: fy, width: fallbackWidth, height: fallbackHeight)
            print("⚠️ cropImage: using fallback cropRectPixels=\(cropRectPixels)")
        }

        if let cropped = cg.cropping(to: cropRectPixels) {
            print("✅ cropImage: cropped cg rect=\(cropRectPixels) cg.size=(\(cg.width),\(cg.height))")
            // Create UIImage from cropped CGImage and normalize orientation to .up so transfers/viewing show correct rotation
            // Use scale 1.0 for the created UIImage to avoid mixing device/backing scales
            var ui = UIImage(cgImage: cropped, scale: 1.0, orientation: .up)
            ui = normalizedImage(ui)

            // Resize to target output resolution (720x1280) while preserving aspect ratio.
            // We scale the cropped image to fill the target area (aspect-fill), then draw it centered
            // into the target canvas so the final image is exactly targetSize but aspect ratio is preserved.
            let targetSize = CGSize(width: 720, height: 1280)
            let srcSize = ui.size

            // Compute scale to fill target (preserve aspect ratio)
            let scaleX = targetSize.width / max(0.0001, srcSize.width)
            let scaleY = targetSize.height / max(0.0001, srcSize.height)
            let scaleToFill = max(scaleX, scaleY)

            let scaledSize = CGSize(width: srcSize.width * scaleToFill, height: srcSize.height * scaleToFill)
            // Center the scaled image in the target canvas (may crop overflow)
            let drawOrigin = CGPoint(x: (targetSize.width - scaledSize.width) / 2.0,
                                     y: (targetSize.height - scaledSize.height) / 2.0)

            // Use a renderer with scale 1.0 (renderer handles backing scale) to avoid
            // issues when mixing cgImage scales on wide-gamut/high-density devices.
            let format = UIGraphicsImageRendererFormat()
            format.scale = 1.0
            format.opaque = true
            if #available(iOS 12.0, *) {
                format.preferredRange = .standard
            }
            print("ℹ️ cropImage: renderer format scale=\(format.scale) opaque=\(format.opaque) preferredRange=\(String(describing: (format as AnyObject).preferredRange))")
            let renderer = UIGraphicsImageRenderer(size: targetSize, format: format)
            let resized = renderer.image { _ in
                ui.draw(in: CGRect(origin: drawOrigin, size: scaledSize))
            }

            self.croppedImage = resized
        }
        #endif
    }

#if canImport(UIKit)
    private func normalizedImage(_ image: UIImage) -> UIImage {
        if image.imageOrientation == .up { return image }
        let size = image.size
        let format = UIGraphicsImageRendererFormat()
        format.scale = image.scale
        let renderer = UIGraphicsImageRenderer(size: size, format: format)
        let rendered = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: size))
        }
        return rendered
    }
#endif
    
    func getImagePosition(in size: CGSize) -> CGPoint {
        let imageWidth = size.width * scale
        let imageHeight = size.height * scale
        
        let x = (size.width - imageWidth) / 2 + offset.width
        let y = (size.height - imageHeight) / 2 + offset.height
        
        return CGPoint(x: x, y: y)
    }
    
    func getImageSize(in size: CGSize) -> CGSize {
        guard let image = selectedImage else { return .zero }
        
        let aspectRatio = image.size.width / image.size.height
        let scaledHeight = size.height * scale
        let scaledWidth = scaledHeight * aspectRatio
        
        return CGSize(width: scaledWidth, height: scaledHeight)
    }
}

File: FlexTargetAndroid/app/src/main/java/com/flextarget/android/ui/imagecrop/ImageCropViewModel.kt
package com.flextarget.android.ui.imagecrop

import android.graphics.Bitmap
import android.graphics.Matrix
import android.graphics.RectF
import androidx.compose.ui.geometry.Size
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class ImageCropViewModel : ViewModel() {
    private val _selectedImage = MutableStateFlow<Bitmap?>(null)
    val selectedImage: StateFlow<Bitmap?> = _selectedImage.asStateFlow()

    private val _croppedImage = MutableStateFlow<Bitmap?>(null)
    val croppedImage: StateFlow<Bitmap?> = _croppedImage.asStateFlow()

    private val _scale = MutableStateFlow(1.0f)
    val scale: StateFlow<Float> = _scale.asStateFlow()

    private val _offset = MutableStateFlow(androidx.compose.ui.geometry.Offset.Zero)
    val offset: StateFlow<androidx.compose.ui.geometry.Offset> = _offset.asStateFlow()

    // Calculated properties
    val minScale: Float = 1.0f
    val maxScale: Float = 5.0f

    fun setSelectedImage(bitmap: Bitmap?) {
        _selectedImage.value = bitmap
        resetTransform()
    }

    fun resetTransform() {
        _scale.value = 1.0f
        _offset.value = androidx.compose.ui.geometry.Offset.Zero
    }

    fun updateScale(newScale: Float) {
        _scale.value = newScale.coerceIn(minScale, maxScale)
    }

    fun updateOffset(newOffset: androidx.compose.ui.geometry.Offset) {
        _offset.value = newOffset
    }

    /// Enforce minimum scale so the image always fills the provided cropSize
    /// and clamp offset so the crop rectangle remains covered by the image.
    fun enforceConstraints(containerSize: Size, cropSize: Size) {
        val bitmap = _selectedImage.value ?: return

        // Ensure the user-visible image (container frame scaled by `scale`) fully
        // covers the crop rectangle `cropSize`. Adjust `scale` if it's too small
        // and clamp the `offset` so the crop rect never reveals background.

        // Compute the minimum user scale needed so the displayed image covers the crop area.
        // displayed size = containerSize * scale, so requiredScale = cropSize / containerSize
        val requiredScaleX = cropSize.width / containerSize.width.coerceAtLeast(0.0001f)
        val requiredScaleY = cropSize.height / containerSize.height.coerceAtLeast(0.0001f)
        val minAllowedScale = kotlin.math.max(1.0f, kotlin.math.max(requiredScaleX, requiredScaleY))

        if (_scale.value < minAllowedScale) {
            _scale.value = minAllowedScale
        }

        // Compute the displayed image size (points) after applying base fill scale and user scale.
        // If we have the image, calculate how the image is scaled to fill the container (scaledToFill semantics):
        //   baseFillScale = max(containerWidth / imageWidth, containerHeight / imageHeight)
        // Displayed image size = image.size * baseFillScale * userScale
        val baseFill = kotlin.math.max(containerSize.width / bitmap.width.toFloat(),
                          containerSize.height / bitmap.height.toFloat())

        val displayedWidth = bitmap.width * baseFill * _scale.value
        val displayedHeight = bitmap.height * baseFill * _scale.value

        // Allowed half-range for the center offset so the crop rect stays fully covered.
        val allowedHalfX = kotlin.math.max(0f, (displayedWidth - cropSize.width) / 2.0f)
        val allowedHalfY = kotlin.math.max(0f, (displayedHeight - cropSize.height) / 2.0f)

        val clampedX = _offset.value.x.coerceIn(-allowedHalfX, allowedHalfX)
        val clampedY = _offset.value.y.coerceIn(-allowedHalfY, allowedHalfY)

        _offset.value = androidx.compose.ui.geometry.Offset(clampedX, clampedY)
    }

    /// Return a clamped offset for a proposed offset without mutating state.
    fun clampedOffset(
        proposed: androidx.compose.ui.geometry.Offset,
        containerSize: Size,
        cropSize: Size,
        scaleOverride: Float? = null
    ): androidx.compose.ui.geometry.Offset {
        val bitmap = _selectedImage.value ?: return proposed
        val effectiveScale = scaleOverride ?: _scale.value

        // Displayed image size (points) = container frame size * effective scale
        // Compute displayed image size (points) using image intrinsic size when possible
        val baseFill = kotlin.math.max(containerSize.width / bitmap.width.toFloat(),
                          containerSize.height / bitmap.height.toFloat())

        val displayedWidth = bitmap.width * baseFill * effectiveScale
        val displayedHeight = bitmap.height * baseFill * effectiveScale

        // Allowed half-range for the center offset so the crop rect remains fully covered
        val allowedHalfX = kotlin.math.max(0f, (displayedWidth - cropSize.width) / 2.0f)
        val allowedHalfY = kotlin.math.max(0f, (displayedHeight - cropSize.height) / 2.0f)

        val clampedX = proposed.x.coerceIn(-allowedHalfX, allowedHalfX)
        val clampedY = proposed.y.coerceIn(-allowedHalfY, allowedHalfY)

        return androidx.compose.ui.geometry.Offset(clampedX, clampedY)
    }

    fun cropImage(cropFrame: androidx.compose.ui.geometry.Rect, canvasSize: Size) {
        val bitmap = _selectedImage.value ?: return

        // Compute baseFillScale using the bitmap size
        val baseFill = kotlin.math.max(canvasSize.width / bitmap.width.toFloat(),
                          canvasSize.height / bitmap.height.toFloat())

        // The displayed image size in points after baseFill and user scale
        val displayedImageSize = Size(
            width = bitmap.width * baseFill * _scale.value,
            height = bitmap.height * baseFill * _scale.value
        )

        // Top-left of the displayed image inside the container (points)
        val imageOrigin = androidx.compose.ui.geometry.Offset(
            x = (canvasSize.width - displayedImageSize.width) / 2.0f + _offset.value.x,
            y = (canvasSize.height - displayedImageSize.height) / 2.0f + _offset.value.y
        )

        // Convert cropFrame (in container points) to bitmap points: (point - imageOrigin) / (baseFill * scale)
        val invScale = 1.0f / (baseFill * _scale.value)

        val imgX = (cropFrame.left - imageOrigin.x) * invScale
        val imgY = (cropFrame.top - imageOrigin.y) * invScale
        val imgW = cropFrame.width * invScale
        val imgH = cropFrame.height * invScale

        // Convert bitmap points to bitmap pixels
        val pxPerPoint = 1.0f // Bitmap coordinates are in pixels
        val originX = (imgX * pxPerPoint).toInt()
        val originY = (imgY * pxPerPoint).toInt()
        val width = (imgW * pxPerPoint).toInt()
        val height = (imgH * pxPerPoint).toInt()

        // Clamp to bitmap bounds
        val clampedOriginX = originX.coerceIn(0, bitmap.width)
        val clampedOriginY = originY.coerceIn(0, bitmap.height)
        val clampedWidth = kotlin.math.min(width, bitmap.width - clampedOriginX)
        val clampedHeight = kotlin.math.min(height, bitmap.height - clampedOriginY)

        // If the resulting rect is empty or too small, attempt a safe fallback
        if (clampedWidth < 1 || clampedHeight < 1) {
            // Fallback: create a centered crop that preserves the target aspect ratio
            // (720x1280) inside the available bitmap bounds.
            val targetSize = Size(720f, 1280f)
            val targetAspect = targetSize.width / targetSize.height

            // Start by trying to use full bitmap height and compute width by aspect
            var fallbackHeight = bitmap.height.toFloat()
            var fallbackWidth = fallbackHeight * targetAspect

            // If that width exceeds bounds, use full width and compute height by aspect
            if (fallbackWidth > bitmap.width) {
                fallbackWidth = bitmap.width.toFloat()
                fallbackHeight = fallbackWidth / targetAspect
            }

            val fx = (bitmap.width - fallbackWidth) / 2.0f
            val fy = (bitmap.height - fallbackHeight) / 2.0f

            val cropped = Bitmap.createBitmap(
                bitmap,
                fx.toInt(),
                fy.toInt(),
                fallbackWidth.toInt(),
                fallbackHeight.toInt()
            )

            // Resize to target output resolution (720x1280)
            val resized = Bitmap.createScaledBitmap(cropped, 720, 1280, true)
            _croppedImage.value = resized
            return
        }

        val cropped = Bitmap.createBitmap(
            bitmap,
            clampedOriginX,
            clampedOriginY,
            clampedWidth,
            clampedHeight
        )

        // Resize to target output resolution (720x1280) while preserving aspect ratio.
        // We scale the cropped image to fill the target area (aspect-fill), then draw it centered
        // into the target canvas so the final image is exactly targetSize but aspect ratio is preserved.
        val targetSize = Size(720f, 1280f)
        val srcSize = Size(cropped.width.toFloat(), cropped.height.toFloat())

        // Compute scale to fill target (preserve aspect ratio)
        val scaleX = targetSize.width / srcSize.width
        val scaleY = targetSize.height / srcSize.height
        val scaleToFill = kotlin.math.max(scaleX, scaleY)

        val scaledSize = Size(
            width = srcSize.width * scaleToFill,
            height = srcSize.height * scaleToFill
        )

        // Center the scaled image in the target canvas (may crop overflow)
        val drawOrigin = androidx.compose.ui.geometry.Offset(
            x = (targetSize.width - scaledSize.width) / 2.0f,
            y = (targetSize.height - scaledSize.height) / 2.0f
        )

        val resized = Bitmap.createBitmap(720, 1280, Bitmap.Config.ARGB_8888)
        val canvas = android.graphics.Canvas(resized)
        val paint = android.graphics.Paint().apply {
            isFilterBitmap = true
            isAntiAlias = true
        }

        canvas.drawBitmap(
            cropped,
            android.graphics.Rect(0, 0, cropped.width, cropped.height),
            android.graphics.RectF(drawOrigin.x, drawOrigin.y, drawOrigin.x + scaledSize.width, drawOrigin.y + scaledSize.height),
            paint
        )

        _croppedImage.value = resized
    }

    fun clearCroppedImage() {
        _croppedImage.value = null
    }

    fun clearSelectedImage() {
        _selectedImage.value = null
        _croppedImage.value = null
        resetTransform()
    }
}